cpp/src/arrow/result.h

```diff
@@ -302,12 +306,26 @@ class Result : public util::EqualityComparable<Result<T>> {
   }
   T operator*() && { return std::move(*this).ValueOrDie(); }
 
-  /// Retrieve the value if ok(), falling back to the provided alternative
-  T ValueOr(T&& alternative) && {
-    if (ok()) {
-      return std::move(*this).ValueOrDie();
+  /// Helper method for implementing Status returning functions in terms of semantically
+  /// equivalent Result returning functions. For example:
+  ///
+  /// Status GetInt(int *out) { return GetInt().Value(out); }
+  template <typename U, typename E = typename std::enable_if<
+                            std::is_constructible<U, T>::value>::type>
+  Status Value(U* out) && {
+    if (!ok()) {
+      return status();
     }
-    return std::move(alternative);
+    *out = U(std::move(*this).ValueOrDie());
+    return Status::OK();
+  }
+
+  /// Move and return the internally stored value or alternative if an error is stored.
+  T ValueOr(T alternative) && {
```

Comment:
`ValueOr` is agnostic about the origin of `alternative`; it could have been copy or move constructed at the call site. If a large vector or similar needs to be passed as `alternative` then at the call site it can be moved into `ValueOr` instead of copied if it isn't needed elsewhere